### ISSUE TITLE
Any plans to open up R_eval_fork to be non blocking?

### ISSUE THREAD

**Issue by jefshe** [jeroen/sys#16]
We currently have a use case where we need to fork a separate R process and monitor its status along with other asynchronous events. 

It would be nice if the fork call and the collecting the result were not part of the same function call so we could monitor the other events similar to out it is implemented in the parallel library. 

I have considered using the parallel library, however, the child process does not exit after completing running which means SIGCHLD is never sent to the parent which is one of the events we listen for

**Comment by jeroen** [comment]
Have you tried the processx pkg? I think @gaborcsardi has covered a similar use case.

**Comment by gaborcsardi** [comment]
Well, `processx` always forks & executes. It does not do fork + eval. 

If that's OK you can use it to monitor background processes.

**Comment by jefshe** [comment]
@jeroen , unfortunately we want to be able to evaluate R expressions and starting a new instance of R would be too slow for our needs as this is the backend for a network application.

I've had a slight refactor of your fork.c to split up the R_eval_fork call into 2 separate functions and it seems to work. Would you accept a PR in that vein? The key difficulty i see is managing multiple children from the same parent but maybe this can just throw an error for the time being

**Comment by jeroen** [comment]
I don't understand your issue with the `parallel` package. If you want the child to quit you can just send a `q('no')` or `tools::pskill(Sys.getpid())` after retrieving the result? Obviouly it cannot quit immediately after completing because it still has to pipe the data to the parent.

I deliberately did not implement the non-blocking fork because `parallel` already does this, plus it has a lot of caveats like non-forksafe calls, race conditions, shared tempdirs, etc. It is rarely the right solution.

**Comment by jefshe** [comment]
without polling for the result in the parallel package, is there a way to know if the r expression has been evaluated?  Our use case is that we want to listen on SIGCHLD and then collect the results of the R proces

**Comment by jeroen** [comment]
I don't think you can collect results after `SIGCHLD` when the process has died? The process needs to be alive to serialize and pipe the result to the parent. Or perhaps I'm not understanding your case.

With parallel you can poll the children via `mccollect(x, wait = FALSE)`.

**Comment by jefshe** [comment]
The parent still has the file descriptor and is still able to read from it even after the child dies.  Your code does something similar already no? The child doesn't wait for the parent to read from the buffer and just raises SIGKILL.

**Comment by jeroen** [comment]
Ah ok if with "result" you only mean status code and not an actual serialized R object.

**Comment by jefshe** [comment]
No I did mean the serialized R object. Maybe I misunderstood you:
```
I don't think you can collect results after SIGCHLD when the process has died? 
```
Did you mean in unix in general or the parallel package? If you are talking about the parallel package, I agree and hence I'm raising this request in your library since you do kill the child process in your package.

If you are talking about unix in general, I tested this out with a simple c program:
```c
#include <stdio.h> 
#include <stdlib.h> 
#include <string.h> 
#include <unistd.h> 
#include <sys/wait.h> 
#include <sys/types.h> 
 
int main(void) 
{ 
        pid_t childpid; 
        char string[] = "Hello, world!\n"; 
        int fd[2]; 
        int status; 
        pipe(fd); 
        if(childpid == 0) 
        { 
                close(fd[0]); 
                write(fd[1], string, (strlen(string)+1)); 
                close(fd[1]); 
                exit(0); 
        } 
        else 
        { 
                wait(&status); 
                char readbuffer[80]; 
                close(fd[1]); 
                read(fd[0], readbuffer, sizeof(readbuffer)); 
                printf("Received string: %s", readbuffer); 
        } 
        return(0); 
}  
```

running strace, you can see the parent happily reads from the pipe after SIGCHLD is raised:
```
wait4(-1, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], 0, NULL) = 2654
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=2654, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
close(4)                                = 0
read(3, "Hello, world!\n\0", 80)        = 15
fstat(1, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
brk(NULL)                               = 0x1718000
brk(0x173a000)                          = 0x173a000
write(1, "Received string: Hello, world!\n", 31) = 31
exit_group(0)                           = ?
+++ exited with 0 +++
Received string: Hello, world!
```

**Comment by jeroen** [comment]
OK but the pipe buffer is max 32kb. You can't serialize an entire object and keep it in the buffer. If you want to send arbitrary R results to the parent, the child needs to be alive to stream pipe it.

**Comment by jefshe** [comment]
that's a good point, I hadn't considered the buffer limit. I guess we can achieve the same thing by waiting on the file descriptor to be written to. Looking at the parallel package, mcparallel does return this so we can use that instead.

Thanks for your help jeroen
